src/hotspot/share/cds/aotClassInitializer.cpp

```diff
@@ -37,62 +37,38 @@
 
 DEBUG_ONLY(InstanceKlass* _aot_init_class = nullptr;)
 
-// Detector for class names we wish to handle specially.
-// It is either an exact string match or a string prefix match.
-class AOTClassInitializer::AllowedSpec {
-  const char* _class_name;
-  bool _is_prefix;
-  int _len;
-public:
-  AllowedSpec(const char* class_name, bool is_prefix = false)
-    : _class_name(class_name), _is_prefix(is_prefix)
-  {
-    _len = (class_name == nullptr) ? 0 : (int)strlen(class_name);
-  }
-  const char* class_name() { return _class_name; }
-
-  bool matches(Symbol* name, int len) {
-    assert(_class_name != nullptr, "caller resp.");
-    if (_is_prefix) {
-      return len >= _len && name->starts_with(_class_name);
-    } else {
-      return len == _len && name->equals(_class_name);
+// Tell if ik is marked as AOT initialization safe via @jdk.internal.vm.annotation.AOTSafeClassInitializer.
+bool AOTClassInitializer::allows_aot_initialization(InstanceKlass* ik) {
```

Comment:
Instead of doing the asserts here, I think the checks should be done in `ClassFileParser`: if a class is marked `@AOTSafeClassInitializer` annotation, then each of its supertypes must either be marked `@AOTSafeClassInitializer`, or have no `<clinit>`. Otherwise an exception should be thrown (ClassFileFormatError??).

This will be a good reminder that if you mark a class as `@AOTSafeClassInitializer`, you should check the `<clinit>` of all of its supertypes.

Then, this function can be removed, and the caller can call `ik->has_aot_safe_initializer()` directly.